microscope/monitor/monitor.py

```diff
@@ -94,29 +97,84 @@ def connect(self):
 
         signal.signal(signal.SIGINT, sigint_in_monitor)
 
+        processor = MonitorOutputProcessor(self.mode)
+
         while resp.is_open():
+            for msg in processor:
+                if msg:
+                    self.queue.put({
+                        'name': self.pod_name,
+                        'node_name': self.node_name,
+                        'output': msg})
+
             resp.update(timeout=1)
             if not self.close_queue.empty():
                 print("Closing monitor")
                 resp.write_stdin('\x03')
                 break
             if resp.peek_stdout():
-                self.queue.put({
-                    'name': self.pod_name,
-                    'node_name': self.node_name,
-                    'output': resp.read_stdout()})
+                processor.add_out(resp.read_stdout())
             if resp.peek_stderr():
-                self.queue.put({
-                    'name': self.pod_name,
-                    'node_name': self.node_name,
-                    'output': resp.read_stderr()})
+                processor.add_err(resp.read_stderr())
 
         resp.close()
 
         self.close_queue.cancel_join_thread()
         self.queue.cancel_join_thread()
 
 
+class MonitorOutputProcessor:
+    def __init__(self, mode: str):
+        self.std_output = queuemodule.Queue()
+        self.std_err = queuemodule.Queue()
+        self.current_msg = ""
+        self.mode = mode
+        self.last_event_wait_timeout = 1500
+        self.last_event_time = 0
+
+    def add_out(self, out: str):
+        for line in out.split("\n"):
+            self.std_output.put(line)
+
+    def add_err(self, err: str):
+        for line in err.split("\n"):
+            self.std_err.put(line)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self) -> str:
+        err = ""
+        while not self.std_err.empty():
+            line = self.std_err.get()
+            err += line + "\n"
+        if err:
+            return err.rstrip("\n")
+
+        prev_event = self.last_event_time
+        while not self.std_output.empty():
+            line = self.std_output.get()
+            if self.mode == "":
+                return line
+            self.last_event_time = int(round(time.time() * 1000))
+
+            if '---' in line:
+                tmp = self.current_msg
+                self.current_msg = line + "\n"
+                return tmp.rstrip("\n")
+            else:
+                self.current_msg += line.strip("\n") + "\n"
+
+        now = int(round(time.time() * 1000))
+        if self.mode != "" and prev_event + self.last_event_wait_timeout < now:
+            if self.current_msg:
+                tmp = self.current_msg
+                self.current_msg = ""
+                return tmp.rstrip("\n")
```

Comment:
Since this is really similar to the handling of the `---` above, it might be clear to have a function that clears and returns current_msg that we call here (and return its return). We can then call is on like 162, assign to tmp, then do line 163. It's redundant but means there's only one way to get the string we emit out of the class.

Reply:
done